### DISCUSSION TITLE
SYNCHRONOUS_DML in a test context - Integration test

### DISCUSSION THREAD

**Discussion by JeroenSfdc** [jongpie/NebulaLogger#454] [Q&A] [answered]
If I'd like to do an integration test with SYNCHRONOUS_DML it fails (where it succeeds with EVENT_BUS).
I tried to get my head around this... This is using the managed package ~~v4.8~~ v4.9. @jongpie 

```
    @IsTest
    static void testNebula_SyncDML() {
        Nebula.Logger.error('Some random error');

        Test.startTest();
        Nebula.Logger.saveLog(Nebula.Logger.SaveMethod.SYNCHRONOUS_DML);
        Test.stopTest();

        Assert.areEqual(1, [SELECT Id FROM Nebula__Log__c].size());
        Assert.areEqual(1, [SELECT Id FROM Nebula__LogEntry__c].size());
    }
```

**Comment by jongpie** [comment, marked as answer]
Hey @JeroenSfdc I think you are doing everything correctly - I think that the issue is unfortunately something that I'll need to address within Nebula Logger's code. Right now, during a test context, Nebula Logger does not load any custom metadata records, including `LoggerSObjectHandler__mdt` records, which are a critical part of how `SYNCHRONOUS_DML` is implemented. This has an unfortunate side effect of making it impossible for you to properly test the `SYNCHRONOUS_DML` save method in the managed package.

I am currently finishing up the next managed package release (`v4.10`) - I'm already behind schedule to get it released, so I don't think I'll be able to address this issue right now. However, improving the testability of Nebula Logger will be a big focus for me over the next few months. My current plan is to...

- Finish release `v4.9.11` of the unlocked package(PR #452) and `v4.10.0` of the managed package - I'm hoping to have these finished within the next week
- Improve the testability of the managed package for release `v4.11.0`, which will be [the Summer '24 release](https://github.com/jongpie/NebulaLogger/milestone/11)

**Reply by jongpie** [reply]
I've logged this as a bug in #457 to track my progress on fixing this issue.

**Reply by JeroenSfdc** [reply]
Thanks @jongpie, it's a minor issue (for us). Thanks for all the amazing work you do around the package.

**Reply by jongpie** [reply]
Awesome, glad to hear that this is only a minor issue for you - I'll keep you updated on my progress over the next few months.

**Reply by jongpie** [reply]
@JeroenSfdc this is finally fixed in [release `v4.15.3`](https://github.com/jongpie/NebulaLogger/releases/tag/v4.15.3) of the unlocked package, and it'll be fixed in the next release of the managed package (`v4.16.0`) in the next few months (tentatively planned for February 2025).
